src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.h

```diff
@@ -0,0 +1,310 @@
+                                    const std::vector<bool> &is_last_chunk);
+
+  // If an utterance is done, we call FinalizeDecoding async on
+  // the threadpool
+  // it will call the utterance's callback when done
+  void FinalizeDecoding(int32 ichannel, CorrelationID corr_id);
+  // static wrapper for thread pool
+  static void FinalizeDecodingWrapper(void *obj, uint64_t ichannel64,
+                                      uint64_t corr_id) {
+    int32 ichannel = static_cast<int32>(ichannel64);
+    static_cast<BatchedThreadedNnet3CudaOnlinePipeline *>(obj)
+        ->FinalizeDecoding(ichannel, corr_id);
+  }
+  // Data members
+
+  BatchedThreadedNnet3CudaOnlinePipelineConfig config_;
+  int32 max_batch_size_;  // extracted from config_
+  // Models
+  const TransitionModel *trans_model_;
+  const nnet3::AmNnetSimple *am_nnet_;
+  std::unique_ptr<OnlineNnet2FeaturePipelineInfo> feature_info_;
+  // Decoder channels currently available, w/ mutex
+  std::vector<int32> available_channels_;
+  std::mutex available_channels_m_;
+
+  // corr_id -> decoder channel map
+  std::unordered_map<CorrelationID, int32> corr_id2channel_;
+
+  const std::vector<int> *fe_threads_channels;
+  const std::vector<SubVector<BaseFloat>> *fe_threads_wave_samples;
+
+  // corr_id -> callback,  w/ mutex
+  // the callback is called once the final lattice is ready
+  std::unordered_map<int,
+                     std::unique_ptr<std::function<void(CompactLattice &)>>>
+      corrid2callbacks_;
+  std::mutex corrid2callbacks_m_;
+
+  // New channels in the current batch. We've just received
+  // their first batch
+  std::vector<int32> list_channels_first_chunk_;
+
+  std::vector<int> n_samples_valid_, n_input_frames_valid_,
+      n_output_frames_valid_;
+
+  // Channels done after current batch. We've just received
+  // their last chunk
+  std::vector<int> list_channels_last_chunk_;
+  std::vector<CorrelationID> list_corr_id_last_chunk_;
+
+  // Feature pipelines, associated to a decoder channel
+  std::vector<std::unique_ptr<OnlineNnet2FeaturePipeline>> feature_pipelines_;
+
+  // Decoder channels used in the curr_batch, double buffering
+  std::vector<int32> decoder_ichannels_, prev_decoder_ichannels_;
+  // Decodables associated with those channels, double buffering
+  std::vector<DecodableCuMatrixMapped *> decoder_decodables_,
+      prev_decoder_decodables_;
+
+  // The decodables are just views into log_posteriors_
+  // List of all decodables possible (one per channel)
+  // If in used in curr_batch, it will be placed in
+  // decoder_decodables_
+  std::vector<DecodableCuMatrixMapped> decodables_;
+
+  // Number of frames already computed in channel (before
+  // curr_batch_)
+  std::vector<int32> channel_frame_offset_;
+
+  // Parameters extracted from the models
+  int input_frames_per_chunk_;
+  int output_frames_per_chunk_;
+  BaseFloat seconds_per_chunk_;
+  BaseFloat samples_per_chunk_;
```

Comment:
Shall we use `int` for variable `samples_per_chunk`?

Reply:
I'll change that. Thanks